### PR TITLE
Update our Review Guidelines

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
...
-* Correct email address used in the commits (if so, travis displays the avatar next to the commit)?
+* Correct email address used in the commits (if so, github displays the avatar next to the commit)?
 * Is this a bugfix, modulesync, breaking change, enhancement, docs update? Label it with `bug`, `modulesync`, `backwards-incompatible`, `enhancement`, `docs`
 * Are updates to the README.md needed but missing? Label it with `needs-docs`
+* Has the file documented params or examples in the header? This needs to be updated as well
 * Are there merge conflicts? Add the `needs-rebase` label
 * Does it need additional tests? Add the `needs-tests` label
-* Are new parameters introduced? They should have datatypes
+* Does it have failing tests? Add the `tests-fail` label
+* Are new parameters introduced? They must datatypes
```
Comment (ekohl) [inline docs/index.md]: s/must/must have/

File: docs/index.md [hunk 2]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
...
 * Are facts used? They should only be accessed via `$facts[]` or [fact()](https://github.com/puppetlabs/puppetlabs-stdlib#fact) from stdlib, but not topscope variables
+* Varibles shouldn't be accessed via topscope: $::$modulename::$param. Instead do: $modulename::$param
```
Comment (ekohl) [inline docs/index.md]: s/$::$/$::/

File: docs/index.md [hunk 3]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
...
 * Are datatypes from stdlib used? Ensure that lowest supported stdlib version is 4.13.1. Check if a newer version introduced the used datatype
-* Are hiera yaml files added for /docs? Ensure that the lowest supported Puppet version is 4.10.9
+* Are hiera yaml files added for data-in-modules? Ensure that the lowest supported Puppet version is 4.10.9
+* Are there new params with datatype Hash or Array? If possible, they should default to empty Hash/Array instead of undef
```
Comment (ekohl) [inline docs/index.md]: I also like stricter data types. ```Array[String]``` over ```Array``` if possible.

File: docs/index.md [hunk 4]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
...
+* Are there new params with datatype Boolean? If possible, they should default to True/False instead of undef
```
Comment (ekohl) [inline docs/index.md]: I'd say ```Optional[Boolean]``` but there are valid cases. A config key not showing up in generated config can be valid.

File: docs/index.md [hunk 5]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
 
 There are a few things that can be checked if you review a pull request against one of our modules:
 
-* Correct email address used in the commits (if so, travis displays the avatar next to the commit)?
+* Correct email address used in the commits (if so, github displays the avatar next to the commit)?
 * Is this a bugfix, modulesync, breaking change, enhancement, docs update? Label it with `bug`, `modulesync`, `backwards-incompatible`, `enhancement`, `docs`
 * Are updates to the README.md needed but missing? Label it with `needs-docs`
+* Has the file documented params or examples in the header? This needs to be updated as well
```
Comment (ekohl) [inline docs/index.md]: Should we start using https://github.com/voxpupuli/puppet-lint-param-docs? It needs a whitelist of files to check (like https://github.com/theforeman/puppet-foreman/blob/645a48c15dbf30901179afb198d4ab392484335d/Rakefile#L21-L24)

File: docs/index.md [hunk 6]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
...
 * Are there merge conflicts? Add the `needs-rebase` label
 * Does it need additional tests? Add the `needs-tests` label
-* Are new parameters introduced? They should have datatypes
+* Does it have failing tests? Add the `tests-fail` label
+* Are new parameters introduced? They must datatypes
 * Are facts used? They should only be accessed via `$facts[]` or [fact()](https://github.com/puppetlabs/puppetlabs-stdlib#fact) from stdlib, but not topscope variables
+* Varibles shouldn't be accessed via topscope: $::$modulename::$param. Instead do: $modulename::$param
 * Are datatypes from stdlib used? Ensure that lowest supported stdlib version is 4.13.1. Check if a newer version introduced the used datatype
-* Are hiera yaml files added for /docs? Ensure that the lowest supported Puppet version is 4.10.9
+* Are hiera yaml files added for data-in-modules? Ensure that the lowest supported Puppet version is 4.10.9
```
Comment (ekohl) [inline docs/index.md]: A rake check would be nice btw.

File: docs/index.md (hunk 1 above)
Comment (bastelfreak) [reply docs/index.md]: fixed

File: docs/index.md (hunk 2 above)
Comment (bastelfreak) [reply docs/index.md]: yeah the example was a bit confusing

File: docs/index.md (hunk 3 above)
Comment (bastelfreak) [reply docs/index.md]: fixed

File: docs/index.md (hunk 4 above)
Comment (bastelfreak) [reply docs/index.md]: correct. I updated the text

File: docs/index.md (hunk 5 above)
Comment (bastelfreak) [reply docs/index.md]: I would like to enforce puppet-strings docs for new params, but most modules don't have existing docs :(. I'm not sure how helpful the linter plugin would be?

File: docs/index.md (hunk 6 above)
Comment (bastelfreak) [reply docs/index.md]: yes :(

File: docs/index.md (hunk 2 above)
Comment (EmilienM) [reply docs/index.md]: s/Varibles/Variables/

Comment (bastelfreak) [reply docs/index.md]: fixed

File: docs/index.md [hunk 7]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
...
+* Does it have failing tests? Add the `tests-fail` label
+* Are new parameters introduced? They must have datatypes
 * Are facts used? They should only be accessed via `$facts[]` or [fact()](https://github.com/puppetlabs/puppetlabs-stdlib#fact) from stdlib, but not topscope variables
+* Varibales shouldn't be accessed via topscope: $::modulename::$param. Instead do: $modulename::$param
 * Are datatypes from stdlib used? Ensure that lowest supported stdlib version is 4.13.1. Check if a newer version introduced the used datatype
-* Are hiera yaml files added for /docs? Ensure that the lowest supported Puppet version is 4.10.9
+* Are hiera yaml files added for data-in-modules? Ensure that the lowest supported Puppet version is 4.10.9
+* Are there new params with datatype Hash or Array? If possible, they should default to empty Hash/Array instead of undef. You can also enforce the datastructure like Array[String]
+* Are there new params with datatype Boolean? If needed, they should default to True/False instead of undef. This is a tricky decision which needs careful reviewing. Sometimes a default is the better approach, sometimes undef
```
Comment (alexjfisher) [inline docs/index.md]: I think there are too many exceptions.  In many cases defaulting to `undef` and then not for example writing a line in a config file is the better option.

File: docs/index.md [hunk 8]
```diff
@@ -115,12 +115,17 @@ Travis will then kick off a build against the new tag created and deploy that bu
...
 * Is this a bugfix, modulesync, breaking change, enhancement, docs update? Label it with `bug`, `modulesync`, `backwards-incompatible`, `enhancement`, `docs`
 * Are updates to the README.md needed but missing? Label it with `needs-docs`
+* Has the file documented params or examples in the header? This needs to be updated as well
 * Are there merge conflicts? Add the `needs-rebase` label
 * Does it need additional tests? Add the `needs-tests` label
-* Are new parameters introduced? They should have datatypes
+* Does it have failing tests? Add the `tests-fail` label
+* Are new parameters introduced? They must have datatypes
 * Are facts used? They should only be accessed via `$facts[]` or [fact()](https://github.com/puppetlabs/puppetlabs-stdlib#fact) from stdlib, but not topscope variables
+* Varibales shouldn't be accessed via topscope: $::modulename::$param. Instead do: $modulename::$param
```
Comment (alexjfisher) [inline docs/index.md]: spelling of 'variables'.  Anyway, is this true?  I thought it was classes that definitely didn't need to be accessed by topscope anymore.  Variables, not sure.

**Comment by alexjfisher** [comment]
Would it be better to suggest amendments to the official style guide?  Then maybe write puppet-lint plugins to enforce the agreed style.

File: docs/index.md (hunk 8 above)
Comment (ekohl) [reply docs/index.md]: Should it be ```$::modulename::param``` rather than ```$param```?

**Comment by bastelfreak** [comment]
@alexjfisher on a long term, yes. But that needs time. I think collecting our opinions in a central place is a good start.

File: docs/index.md (hunk 7 above)
Comment (bastelfreak) [reply docs/index.md]: My phrasing isn't perfect. I will try to rephrase it.

Comment (bastelfreak) [reply docs/index.md]: updated it

File: docs/index.md (hunk 8 above)
Comment (bastelfreak) [reply docs/index.md]: What I wanted to say is: You shouldn't use $:: anymore, no matter if it's a class, param or fact.

Comment (bastelfreak) [reply docs/index.md]: updated it

Comment (alexjfisher) [reply docs/index.md]: Are you sure?  I thought it was just classes where this was 100% true.  and what about parameters from the ENC?

Comment (ekohl) [reply docs/index.md]: Aren't parameters from the ENC presented as facts? Haven't checked this tbh but I see them in the same way.

Comment (rnelson0) [reply docs/index.md]: There is no dynamic scoping of variables or class names anymore.

**Pull request merged**